### ISSUE TITLE
Windows GVim :GoCallers Problem

### ISSUE THREAD

**Issue by umitkablan** [fatih/vim-go#872]
Hi @fatih,

I came up with another path backslash problem and a toy/raw solution for :GoCallers command.

```
diff --git a/autoload/go/package.vim b/autoload/go/package.vim
index 4f75613..f577da4 100644
--- a/autoload/go/package.vim
+++ b/autoload/go/package.vim
@@ -59,7 +59,7 @@ function! go#package#ImportPath(arg)
     let dirs = go#package#Paths()
 
     for dir in dirs
-        if len(dir) && match(path, dir) == 0
+        if len(dir) && matchstr(escape(path, '\/'), escape(dir, '\/')) == 0
             let workspace = dir
         endif
     endfor
@@ -68,8 +68,8 @@ function! go#package#ImportPath(arg)
         return -1
     endif
 
-    let srcdir = substitute(workspace . '/src/', '//', '/', '')
-    return substitute(path, srcdir, '', '')
+    let srcdir = substitute(workspace . '\src\', '//', '/', '')
+    return path[len(srcdir):]
 endfunction
 
 function! go#package#FromPath(arg)
```

I wish it will help you to solve the issue in a more elegant way.

Best,

**Comment by fatih** [comment]
Thanks @umitkablan for the report. What exactly is this solving? I assume GoCallers is not working well because it can't return correct paths?

**Comment by HakShak** [comment]
Basically you get the "-1" problem because it can't match anything due to the escape sequence and the slash direction.

How would I "log" the results of this function? (Never done any vim plugin work)

**Comment by umitkablan** [comment]
Hi @fatih,
Sorry for the late response - just saw your comment.

When I do :GoCallers  I get
    vim-go: current directory is not inside of a valid GOPATH

And after digging inside I sketched out such a solution to help you reason about the solution.

Actually as far as I used today, you had vimproc awareness in the plugin and thought you are going to that direction in order to avoid such platform details. But it still happens after vimproc and my vimproc is working.

**Comment by umitkablan** [comment]
And I experimented more with newly added vimproc-utilized system runs.
When I have let `g:go#use_vimproc = 0` in my vimrc the above patch is enough to GoCallers successfully. But when I `let g:go#use_vimproc = 1` enabled then I need additional patch:
```
--- a/autoload/go/guru.vim
+++ b/autoload/go/guru.vim
@@ -6,8 +6,9 @@ func! s:RunGuru(mode, format, selected, needs_scope) range abort
     if empty(bin_path)
         return {'err': "bin path not found"}
     endif
+    let bin_path = shellescape(fnamemodify(bin_path, ':p:gs?\\?/?'))
 
-    let filename = expand('%:p')
+    let filename = fnamemodify(expand('%:p'), ':p:gs?\\?/?')
     let dirname = expand('%:p:h')
     let pkg = go#package#ImportPath(dirname)
```
Which makes me think vimproc needs forward slashes to work correctly independent of OS.

**Comment by fatih** [comment]
@umitkablan 

Actually vimproc is just a plugin that allows us to call process in async way. It's not meant for cross compatibility. I'm going to remove it in the future with vim 8.0.

I've added one your fix here: https://github.com/fatih/vim-go/pull/914  Can you try it and see if this works for you? I've didn't changed the bin_path as I think we don't need it, but if that fixes l'll can add it. If you wish to include the patch yourself, I can abandon my PR and merge yours, just let me know please.

**Comment by fatih** [comment]
Ping @umitkablan @HakShak

**Comment by umitkablan** [comment]
Hi @fatih,

I can, in short, say this patch is not working and I can prepare a PR if the plan is changing to/from forward/backward slashes? This PR could cover all filepath tricks needed for Windows. Are we gonna concentrate on that slash changing strategy or something else?

**Comment by umitkablan** [comment]
Hi @fatih,

My above observation was with:

`let g:go#use_vimproc = 1`

And it seems we need just this patch when we do not use vimproc@vim-go on windows platform:

```
diff --git a/autoload/go/package.vim b/autoload/go/package.vim
index 2aefd5d..023dd96 100644
--- a/autoload/go/package.vim
+++ b/autoload/go/package.vim
@@ -59,7 +59,7 @@ function! go#package#ImportPath(arg)
   let dirs = go#package#Paths()
 
   for dir in dirs
-    if len(dir) && match(path, dir) == 0
+    if len(dir) && matchstr(escape(path, '\/'), escape(dir, '\/')) == 0
       let workspace = dir
     endif
   endfor
@@ -68,8 +68,8 @@ function! go#package#ImportPath(arg)
     return -1
   endif
 
-  let srcdir = substitute(workspace . '/src/', '//', '/', '')
-  return substitute(path, srcdir, '', '')
+  let srcdir = substitute(workspace . '\src\', '//', '/', '')
+  return path[len(srcdir):]
 endfunction
 
 function! go#package#FromPath(arg)

```

So, I believe we can fix issue without considering vimproc since it complicates stuff - it needs forward slashes and the commands like guru.exe do work with backward slashes... In order to avoid complexity at the moment I will not use vimproc within vim-go and use above patch to fix GoCallers.

**Comment by HakShak** [comment]
Mobile with no Windows machine at the moment. Will chime in once I get a chance to test.

**Comment by fatih** [comment]
@umitkablan is there any chance you can open a PR, and some windows users, such as @HakShak can test it? Unfortunately I'm helpless here as I don't have a way to provide or to test your fix.

**Comment by umitkablan** [comment]
Hi @fatih,

Sure I can send a PR regarding my change - it would be easier to comment/check the diff there. Since I am not working 100% in a Windows platform I couldn't check, sorry.

**Comment by fatih** [comment]
oops, I thought you're using Windows, or not ? You opened the issue, I'm confused now 🤔

**Comment by umitkablan** [comment]
I use - not all the time, that's it..

**Comment by fatih** [comment]
Issue fixed with #999. Thanks @umitkablan
